subprojects/workers/src/main/java/org/gradle/workers/internal/InProcessCompilerDaemonFactory.java

```diff
@@ -0,0 +1,144 @@
+        this.classLoaderFactory = classLoaderFactory;
+        this.buildOperationWorkerRegistry = buildOperationWorkerRegistry;
+        this.buildOperationExecutor = buildOperationExecutor;
+    }
+
+    @Override
+    public WorkerDaemon getDaemon(final Class<? extends WorkerDaemonProtocol> serverImplementationClass, File workingDir, final DaemonForkOptions forkOptions) {
+        return new WorkerDaemon() {
+            @Override
+            public <T extends WorkSpec> DefaultWorkResult execute(WorkerDaemonAction<T> action, T spec) {
+                return execute(action, spec, buildOperationWorkerRegistry.getCurrent(), buildOperationExecutor.getCurrentOperation());
+            }
+
+            @Override
+            public <T extends WorkSpec> DefaultWorkResult execute(final WorkerDaemonAction<T> action, final T spec, Operation parentWorkerOperation, BuildOperationExecutor.Operation parentBuildOperation) {
+                BuildOperationWorkerRegistry.Completion workerLease = parentWorkerOperation.operationStart();
+                BuildOperationDetails buildOperation = BuildOperationDetails.displayName(action.getDescription()).parent(parentBuildOperation).build();
+                try {
+                    return buildOperationExecutor.run(buildOperation, new Transformer<DefaultWorkResult, BuildOperationContext>() {
+                        @Override
+                        public DefaultWorkResult transform(BuildOperationContext buildOperationContext) {
+                            return executeInWorkerClassLoader(serverImplementationClass, action, spec, forkOptions.getClasspath(), forkOptions.getSharedPackages());
+                        }
+                    });
+                } finally {
+                    workerLease.operationFinish();
+                }
+            }
+        };
+    }
+
+    private <T extends WorkSpec> DefaultWorkResult executeInWorkerClassLoader(Class<? extends WorkerDaemonProtocol> serverImplementationClass, WorkerDaemonAction<T> action, T spec, Iterable<File> classpath, Iterable<String> sharedPackages) {
+        ClassLoader workerClassLoader = createWorkerClassLoader(serverImplementationClass, classpath, sharedPackages);
+        try {
+            LOGGER.info("Executing {} in in-process worker.", action.getDescription());
+            Callable<?> worker = transferWorkerIntoWorkerClassloader(action, spec, workerClassLoader);
+            Object result = worker.call();
+            DefaultWorkResult workResult = transferResultFromWorkerClassLoader(result);
+            LOGGER.info("Successfully executed {} in in-process worker.", action.getDescription());
+            return workResult;
+        } catch (Exception e) {
+            LOGGER.info("Exception executing {} in in-process worker: {}.", action.getDescription(), e);
```

Comment:
I don't know that we need a logging statement for this - the exception will bubble up and either be handled or end up in the error output.